### PR TITLE
Add a utility function to get all of the PartitionedFile for an ExecutionPlan

### LINKED ISSUE #5566: Add a utility function to collect all of the PartitionedFile which is needed for an ExecutionPlan
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently the `TreeNodeRewriter` is as a visitor to transform a node to another. However, sometimes we don't need to do the transformation and what we want is only to collect some info from the node. To achieve this, it's better to introduce another visitor for collecting info and keep the node unchanged.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #5566.

# Rationale for this change

Currently the `TreeNodeRewriter` is as a visitor to transform a node to another. However, sometimes we don't need to do the transformation and what we want is only to collect some info from the node. To achieve this, it's better to introduce another visitor for collecting info and keep the node unchanged.

# What changes are included in this PR?

- Introduced a common tree node visitor trait
- Implemented the visitable trait for `ExecutionPlan`
- Introduced a utility function to get all of the PartitionedFile for an ExecutionPlan based on the visitor mentioned above.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @mingmwang, @alamb, could you help review this PR?

**Comment by mingmwang** [comment]
I think you can  move the `TreeNodeRewriter` to the new added `tree_node.rs` also.

**Comment by alamb** [comment]
@mustafasrepo  do you have time to review this PR?

**Comment by yahoNanJing** [comment]
> I think you can move the TreeNodeRewriter to the new added tree_node.rs also.

I'll add another commit to this PR for this.

**Comment by mustafasrepo** [comment]
> @mustafasrepo do you have time to review this PR?

Sure. I will look at it.

**Comment by yahoNanJing** [comment]
Unfortunately, due to the Orphan rule, I fail to combine the trait `TreeNodeRewriter` for both of the `ExecutionPan` and `PhysicalExpr`

File: datafusion/core/src/physical_plan/tree_node/rewritable.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// under the License.
+
+//! Tree node rewritable implementations
+
+use crate::physical_plan::tree_node::TreeNodeRewritable;
+use crate::physical_plan::{with_new_children_if_necessary, ExecutionPlan};
+use datafusion_common::Result;
+use std::sync::Arc;
+
+impl TreeNodeRewritable for Arc<dyn ExecutionPlan> {
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/tree_node/rewritable.rs:25]: Is there any reason for implementing `TreeNodeRewritable` trait in another file. I guess, you could have implement this trait in the file where `ExecutionPlan` is implemented. I am not familiar with these pattern. I am asking to understand better.

File: datafusion/core/src/physical_plan/tree_node/visitable.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+//! Tree node visitable implementations
+
+use crate::physical_plan::tree_node::TreeNodeVisitable;
+use crate::physical_plan::ExecutionPlan;
+use std::sync::Arc;
+
+impl TreeNodeVisitable for Arc<dyn ExecutionPlan> {
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/tree_node/visitable.rs:24]: Similar to the above case I guess `TreeNodeVisitable` could have been implemented where `ExecutionPlan` is defined. Is there any reason for not doing so.

File: datafusion/core/src/physical_plan/tree_node/rewritable.rs (hunk 1 above)
Comment (yahoNanJing) [reply datafusion/core/src/physical_plan/tree_node/rewritable.rs:25]: If the Orphan rule is relaxed in the future, the same trait of `TreeNodeRewritable` can be used for the `PhysicalExpr`. The traits defined in the tree_node file are common abstractions. They are not just for `ExecutionPlan`

**Comment by yahoNanJing** [comment]
Thanks @mustafasrepo. I'll merge this PR.

**Comment by mingmwang** [comment]
@yahoNanJing 
I think if we just to collect/search some information from a tree like structure, we do not need to define a Visitor `Trait`.
A visitor in this case is actually just a map function (closure) we want to apply to a tree structure. 
For example, if we want to traverse a Vec/List container and collect/search some information, we do not need to define a `Visitor`.

**Comment by yahoNanJing** [comment]
Thanks @mingmwang for your suggestion. I'll raise another PR to replace the Visitor by closure which will be much easier to use.

File: datafusion/core/src/physical_plan/tree_node/rewritable.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_plan/tree_node/rewritable.rs:25]: I understand why `TreeNodeRewritable` trait is defined in `datafusion/core/src/physical_plan/tree_node/mod.rs`. What I mean was that `Arc<dyn ExecutionPlan>` implements `TreeNodeRewritable` trait. I guess, we could do this implementation in `datafusion/core/src/physical_plan/mod.rs`. However, it is just a stylistic issue. I just wondered the reason why you did it in  `datafusion/core/src/physical_plan/tree_node/rewritable.rs`

**Pull request merged**